circuit/collections/src/sparse_kary_merkle_tree/helpers/path_hash.rs

```diff
@@ -0,0 +1,266 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+use snarkvm_circuit_algorithms::{BHP, Hash, Keccak, Poseidon};
+
+/// A trait for a Merkle path hash function.
+pub trait PathHash<E: Environment> {
+    type Hash: Clone
+        + Default
+        + Inject<Primitive = <Self::Primitive as console::sparse_kary_merkle_tree::PathHash>::Hash>
+        + Eject<Primitive = <Self::Primitive as console::sparse_kary_merkle_tree::PathHash>::Hash>
+        + Equal<Output = Boolean<E>>
+        + Ternary<Boolean = Boolean<E>, Output = Self::Hash>;
+    type Primitive: console::sparse_kary_merkle_tree::PathHash;
+
+    /// Returns the hash of the given child nodes.
+    fn hash_children(&self, children: &[Self::Hash]) -> Self::Hash;
+
+    /// Returns the empty hash.
+    fn hash_empty<const ARITY: u8>(&self) -> Self::Hash {
+        let children = vec![Self::Hash::default(); ARITY as usize];
+        self.hash_children(&children)
+    }
+}
+
+impl<E: Environment, const NUM_WINDOWS: u8, const WINDOW_SIZE: u8> PathHash<E> for BHP<E, NUM_WINDOWS, WINDOW_SIZE> {
+    type Hash = Field<E>;
+    type Primitive = console::algorithms::BHP<E::Network, NUM_WINDOWS, WINDOW_SIZE>;
+
+    /// Returns the hash of the given child nodes.
+    fn hash_children(&self, children: &[Self::Hash]) -> Self::Hash {
+        let mut input = Vec::new();
+        // Prepend the nodes with a `false` & `true` bit.
+        input.push(Boolean::constant(false));
+        input.push(Boolean::constant(true));
+        for child in children {
+            child.write_bits_le(&mut input);
+        }
+        // Hash the input.
+        Hash::hash(self, &input)
+    }
+}
+
+impl<E: Environment, const RATE: usize> PathHash<E> for Poseidon<E, RATE> {
+    type Hash = Field<E>;
+    type Primitive = console::algorithms::Poseidon<E::Network, RATE>;
+
+    /// Returns the hash of the given child nodes.
+    fn hash_children(&self, children: &[Self::Hash]) -> Self::Hash {
+        let mut input = Vec::with_capacity(1 + children.len());
+        // Prepend the nodes with a `1field` byte.
+        input.push(Self::Hash::one());
+        for child in children {
+            input.push(child.clone());
+        }
+        // Hash the input.
+        Hash::hash(self, &input)
+    }
+}
+
+impl<E: Environment, const TYPE: u8, const VARIANT: usize> PathHash<E> for Keccak<E, TYPE, VARIANT> {
+    type Hash = BooleanHash<E, VARIANT>;
+    type Primitive = console::algorithms::Keccak<TYPE, VARIANT>;
+
+    /// Returns the hash of the given child nodes.
+    fn hash_children(&self, children: &[Self::Hash]) -> Self::Hash {
+        let mut input = Vec::new();
+        // Prepend the nodes with a `false` & `true` bit.
+        input.push(Boolean::constant(false));
```

Comment:
Nit: Consider making the magic constants actual constants, e.g. `PREFIX_BITS_SMT_KECCAK: [bool; 2];`. The snarkVM synthesizer uses this in a few places and it makes it much harder for a future idiot to mess up.